R/glossary.R

```diff
@@ -0,0 +1,390 @@
+    names(drop) <- rep("*", length(drop))
+    warn("Dropping duplicate glossary labels:", body = drop)
+    x <- x[!duplicates]
+  }
+  type_chr <- vapply(x, inherits, what = "character", FUN.VALUE = TRUE)
+  len <- vapply(x, length, FUN.VALUE = 1)
+  if(!all(type_chr & len==1)) {
+    abort("Every `glossary` entry must have type character and length 1.")
+  }
+  glossary <- Map(x, names(x), f = function(def, label) {
+    data <- c(definition = def, abbreviation = label)
+    as_glossary_entry(data)
+  })
+  class(glossary) <- c("tex_glossary", "glossary", "list")
+  glossary
+}
+
+as_glossary_entry <- function(x) {
+  x <- as.list(x)
+  class(x) <- c("glossary_entry", "list")
+  x
+}
+
+abort_bad_glo_labels <- function(x, what) {
+  if(!all(what %in% names(x))) {
+    bad <- setdiff(what, names(x))
+    names(bad) <- rep("x", length(bad))
+    msg <- c("Requested definitions not found in glossary object", bad)
+    abort(msg)
+  }
+}
+
+#' Return formatted notes from a tex glossary file or glossary object
+#'
+#' @param x path to a tex glossary file, a glossary object, or a list that can
+#' be coerced to a glossary object with [as_glossary()].
+#'
+#' @inheritParams read_glossary
+#' @inheritParams st_notes_glo
+#'
+#' @examples
+#' file <- system.file("glo", "glossary.tex", package = "pmtables")
+#'
+#' glossary_notes(file, WT, CRCL)
+#'
+#' g <- as_glossary(list(ss = "steady state", ALB = "albumin", WT = "weight"))
+#'
+#' glossary_notes(g, ALB, ss)
+#'
+#' @seealso [st_notes_glo()], [read_glossary()]
+#' @export
+glossary_notes <- function(x, ...) UseMethod("glossary_notes")
+
+#' @param ... unquoted glossary labels to be included in the notes text.
+#' @param format passed to [read_glossary()].
+#' @rdname glossary_notes
+#' @export
+glossary_notes.character <- function(x, ..., format = guess_glo_fmt(x)) {
+  x <- read_glossary(x, format = format)
+  glossary_notes(x, ...)
+}
+
+#' @rdname glossary_notes
+#' @export
+glossary_notes.list <- function(x, format = guess_glo_fmt(x), ...) {
+  x <- as_glossary(x, format = format)
+  glossary_notes(x, ...)
+}
+
+#' @rdname glossary_notes
+#' @export
+glossary_notes.glossary <- function(x, ..., sep = ": ", collapse = "; ",
+                                    labels = NULL) {
+  labels <- cvec_cs(labels)
+  dots <- eval_select(expr(c(...)), x)
+  labels <- c(names(dots), labels)
+  if(!length(labels)) labels <- names(x)
+  build_glossary_notes(x, labels, sep, collapse)
+}
+
+build_glossary_notes <- function(glossary, labels, sep, collapse) {
+  abort_bad_glo_labels(glossary, labels)
+  glossary <- glossary[labels]
+  abb <- vapply(glossary, "[[", "abbreviation", FUN.VALUE = "a")
+  notes <- lapply(glossary, "[[", "definition")
+  notes <- paste0(abb, sep, notes)
+  if(is.character(collapse)) {
+    notes <- paste0(notes, collapse = collapse)
+  }
+  notes
+}
+
+#' @export
+print.glossary_entry <- function(x, ...) {
+  cat(ans <- paste0(x$definition, " (", x$abbreviation, ")"))
```

Comment:
Nit-pick: it'd be good to include a newline at the end of this `cat` to play nicely with consoles other than RStudio that don't automatically tack on a missing newline. For example, with plain R from a terminal, the `>` gets stuck to the end of the output:

```
> as_glossary(list(ss = "steady state"))[[1]]
steady state (ss)>
```

Reply:
Never thought of this before.